### PR TITLE
Redis: improvements

### LINKED ISSUE #1613: RedisEmbeddingStore: Supporting different index types for metadata in RedisEmbeddingStore
### Discussed in https://github.com/langchain4j/langchain4j/discussions/1612

<div type='discussions-op-text'>

<sup>Originally posted by **mortensi** August 20, 2024</sup>
Ingesting data in a RedisEmbeddingStore with metadata creates a TEXT index. This is useless for non-textual metadata (like saying you can't have numeric metadata, exact match TAG indexes, or geographical locations).

```
"FT.CREATE" "movies_20240820_230935_idx" "ON" "JSON" "PREFIX" "1" "embedding:" "SCHEMA" "$.text" "AS" "text" "TEXT" "WEIGHT" "1.0" "$.vector" "AS" "vector" "VECTOR" "HNSW" "8" "DIM" "1536" "DISTANCE_METRIC" "COSINE" "TYPE" "FLOAT32" "INITIAL_CAP" "5" "$.score" "AS" "score" "TEXT" "WEIGHT" "1.0" "$.date_x" "AS" "date_x" "TEXT" "WEIGHT" "1.0"
```

Supporting schemas in the same way as LangChain for Python does is nice to have.

    index_schema = {
        "tag": [{"name": "genre"},
                {"name": "country"}],
        "text": [{"name": "names"}],
        "numeric": [{"name": "revenue"},
                    {"name": "score"},
                    {"name": "date_x"}]
    }
    
Redis.from_texts(index_schema=index_schema, ...)

Metadata indexed with the right type of index enable advanced search and RAG use cases.</div>

### PR DESCRIPTION
**Pull request by Martin7-1**
## Issue

Closes #1613 and #1618 

## Change

1. Make `dimension` parameter optional in ctor if index is exist.
2. Support different index type in metadata creation.
3. Upgrade `jedis` to 5.1.5

This PR should be merged after #1340 fixed.

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

## Checklist for changing existing embedding store integration

- [x] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by Martin7-1** [comment]
image

All tests are green in my local environment. But some of them are failed in Github Action.

**Comment by Martin7-1** [comment]
image

All tests are green in my local environment. But some of them are failed in Github Action.

**Comment by langchain4j** [comment]
`should_add_embedding_with_segment_with_metadata` fails for me with:

```
java.lang.NumberFormatException: For input string: "9.22337203685e+18"

	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Long.parseLong(Long.java:708)
	at java.base/java.lang.Long.parseLong(Long.java:831)
	at dev.langchain4j.data.document.Metadata.getLong(Metadata.java:201)
	at dev.langchain4j.store.embedding.EmbeddingStoreIT.should_add_embedding_with_segment_with_metadata(EmbeddingStoreIT.java:65)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)


```

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java [hunk 1]
```diff
@@ -64,8 +66,14 @@ SchemaField[] toSchemaFields() {
                 .build());
 
         if (metadataKeys != null) {
+            metadataKeys.forEach(metadataKey -> {
+                if (!schemaFieldMap.containsKey(metadataKey)) {
+                    throw new RuntimeException("Schema field " + metadataKey + " not found");
+                }
+            });
             for (String metadataKey : metadataKeys) {
-                fields.add(TextField.of(JSON_PATH_PREFIX + metadataKey).as(metadataKey).weight(1.0));
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java:79]: what about keeping the old logic if `metadataKeys` was not provided by the user?

**Review by ghost — COMMENTED** [review]
@Martin7-1 thank you!

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java (hunk 1 above)
Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisSchema.java:79]: Nice point. Already done.

**Comment by Martin7-1** [comment]
> `should_add_embedding_with_segment_with_metadata` fails for me with:
> 
> ```
> java.lang.NumberFormatException: For input string: "9.22337203685e+18"
> 
> 	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
> 	at java.base/java.lang.Long.parseLong(Long.java:708)
> 	at java.base/java.lang.Long.parseLong(Long.java:831)
> 	at dev.langchain4j.data.document.Metadata.getLong(Metadata.java:201)
> 	at dev.langchain4j.store.embedding.EmbeddingStoreIT.should_add_embedding_with_segment_with_metadata(EmbeddingStoreIT.java:65)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> ```

@langchain4j Done. It seems that the Redis `NumericType` range is different from the Java range, so I'm treating Long and Double as `TextField`.

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 2]
```diff
@@ -55,19 +57,21 @@ public RedisEmbeddingStore(String host,
...
         ensureNotBlank(host, "host");
         ensureNotNull(port, "port");
-        ensureNotNull(dimension, "dimension");
 
         this.client = user == null ? new JedisPooled(host, port) : new JedisPooled(host, port, user, password);
         this.schema = RedisSchema.builder()
                 .indexName(getOrDefault(indexName, "embedding-index"))
                 .dimension(dimension)
                 .metadataKeys(metadataKeys)
+                .schemaFieldMap(schemaFieldMap)
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]:
```suggestion
                .schemaFieldMap(copyIfNotNull(schemaFieldMap))
```

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 3]
```diff
@@ -295,8 +310,16 @@ public Builder metadataKeys(Collection<String> metadataKeys) {
             return this;
         }
 
+        /**
+         * @param schemaFieldMap Metadata schemaField, the key set must be the same as {@link Builder#metadataKeys(Collection)} (optional)
+         */
+        public Builder schemaFiledMap(Map<String, SchemaField> schemaFieldMap) {
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]:
```suggestion
        public Builder metadataKeys(Map<String, SchemaField> metadataKeys) {
```
It would probably feel a bit strange for the user to provide the same information (metadata keys) twice. And there will be a chance that `metadataKeys` and `schemaFiledMap` are not in sync.

What about having a `metadataKeys` method instead where user can specify not only the keys but also types?

So this:
```java
RedisEmbeddingStore.builder()
                ...
                .metadataKeys(Set.of("one", "two"))
                .schemaFiledMap(Map.of("one", NumericField.of(...), "two", NumericField.of(...)))
                .build();
```
Will be simplified to this:
```java
RedisEmbeddingStore.builder()
                ...
                .schemaFiledMap(Map.of("one", NumericField.of(...), "two", NumericField.of(...)))
                .build();
```

File: langchain4j-redis/src/test/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStoreIT.java [hunk 4]
```diff
@@ -56,4 +71,9 @@ protected EmbeddingStore<TextSegment> embeddingStore() {
     protected EmbeddingModel embeddingModel() {
         return embeddingModel;
     }
+
+    @Override
+    protected void ensureStoreIsReady() {
```
Comment (langchain4j) [inline langchain4j-redis/src/test/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStoreIT.java]: sorry for that, this can be removed now (default implementation is now empty)

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 5]
```diff
@@ -169,9 +173,19 @@ private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<T
...
+                    // NOTE: do not check metadata key is included in RedisSchema#metadataKeys
                     fields.put(schema.scalarFieldName(), textSegment.text());
-                    fields.putAll(textSegment.metadata().asMap());
+                    Map<String, Object> metadataMap = textSegment.metadata().toMap();
+
+                    // NOTE: since Java types and Redis types don't always correspond, there is a conversion to be done here.
+                    for (Map.Entry<String, SchemaField> fieldEntry : schema.schemaFieldMap().entrySet()) {
+                        String metadataKey = fieldEntry.getKey();
+                        SchemaField field = fieldEntry.getValue();
+                        if (field instanceof TextField && metadataMap.containsKey(metadataKey)) {
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Could you please provide an example when this conversion is necessary?

**Review by ghost — COMMENTED** [review]
@Martin7-1 thanks a lot!

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java (hunk 5 above)
Comment (langchain4j) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Is it because of this?

> It seems that the Redis NumericType range is different from the Java range, so I'm treating Long and Double as TextField.

Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: > Is it because of this?
> 
> > It seems that the Redis NumericType range is different from the Java range, so I'm treating Long and Double as TextField.

Yes. When I treat `Long` and `Double` as `NumericType` in Redis, it seems that `Redis` can't handle it in a right way, and the exception will occur:

```java
java.lang.NumberFormatException: For input string: "9.22337203685e+18"

	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Long.parseLong(Long.java:708)
	at java.base/java.lang.Long.parseLong(Long.java:831)
	at dev.langchain4j.data.document.Metadata.getLong(Metadata.java:201)
```

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java (hunk 3 above)
Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: @langchain4j Done. I make `metadatakeys` deprecated and will use `TextField` as default SchemaField if users still use this method.

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java (hunk 2 above)
Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Done.

**Comment by Martin7-1** [comment]
Also, I observe a platform difference. When I run IT on MacOS M1, all tests are green (`Long` metadata can store correctly on Redis), but when I run IT on Windows, the exception occur.

Now, I think we should treat all number correctly, users can add Long and Double as `NumericType`, but in our IT, I add `supportLong()` in `EmbeddingStoreIT` in order to pass all tests. Anyway, I think the corner case won't affect users very much, as it's not our responsibility to ensure the range of number the vector store can store.

**Comment by Martin7-1** [comment]
@langchain4j Hi! I found the issue is related to [client.ft.search() always returning strings](https://github.com/redis/node-redis/discussions/2622#discussioncomment-9821650). The "9.22337203685e+18" string occur because all fields are returned as `String` type.

Now I fix it :)

Currently I'm working on Filter with Redis（I will open a new issue and new PR）.

**Comment by Martin7-1** [comment]
@langchain4j Hi, this PR is ready to review as #1981 is merged now! This PR should be merged then I can migrate `Redis` to new `EmbeddingStore` search API. (see #1833 )

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java (hunk 3 above)
Comment (langchain4j) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: I am not sure about `schemaFiledMap` method name, perharps we should keep `metadataKeys` or something like `metadataFields`? It sohuld have "metadata" in the name for sure

**Review by ghost — COMMENTED** [review]
@Martin7-1 thank you!

**Comment by Martin7-1** [comment]
> I am not sure about schemaFiledMap method name, perharps we should keep metadataKeys or something like metadataFields? It sohuld have "metadata" in the name for sure

@langchain4j What about just `metadataMap` or `metadataTypeMap`? Actually it is mapping metadata keys to metadata type.

**Comment by langchain4j** [comment]
@Martin7-1 `metadataConfig` perharps? Why do we need `Map` at the end?

**Comment by Martin7-1** [comment]
> `metadataConfig` perharps? Why do we need `Map` at the end?

Sure. Adding `Map` at the end is just a personal habit. :)

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java [hunk 6]
```diff
@@ -314,20 +317,29 @@ public Builder dimension(Integer dimension) {
...
-            this.metadataKeys = metadataFieldsName;
-            return this;
+            return metadataKeys(metadataFieldsName);
         }
 
         /**
          * @param metadataKeys Metadata keys that should be persisted (optional)
+         * @deprecated use {@link #metadataConfig(Map)}} instead
          */
+        @Deprecated
```
Comment (langchain4j) [inline langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Should we really deprecate this? What if users are ok with default metadata config?

Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Currently I implement this deprecated method by default (like previous set all metadata field as `TextField`). Actually I don't have plan to remove it right now, but using `metadataConfig()` will be more concrete and can use filter then.

I can remove deprecate if you think we should keep it.

Comment (langchain4j) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: My concern is that by deprecating this method we will be forcing users to explicitly specify types for all metadata keys. I am not sure if it is good or not, WDYT? If user just wants to store metadata values, they should be able to do so without the need to specify types explicitly, or?

Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: hmmm, you are right. What about splitting it into like `tagFields(Collection<String> keys)`, `textFields(Collection<String> keys)` and so on. Thus users can just specify the metadata keys.

If users want to customize the field setting, they should call `metadataConfig` and pass types.

In ctor, we can merge all fields passed by `metadataConfig`, `tagFields`, `textFields` and so on, and then deduplicate it by name. (priority: metadataConfig > others).

WDYT?

Comment (langchain4j) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Sorry, but what are `tagFields`?

Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Tag type is a field type represents shorter string. see [Redis Search field and type options](https://redis.io/docs/latest/develop/interact/search-and-query/basic-constructs/field-and-type-options/#tag-fields).

What I mean is that represent convinient builder method for all type, and take `Collection<String> keys` as param. Users can just pass the name of keys and don't need to specify field type. Some field type may has some advaned setting (such as `TextField` has `weight` param), users can specify it by passing `metadataConfig`.

**Comment by Martin7-1** [comment]
About Codacy Static Code Analysis, it says `Avoid throwing raw exception types`. But I think introducing a new exception which just wrap `RuntimeException` is unnecessary...

Any good solution?

File: langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java (hunk 6 above)
Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: @langchain4j Hi! Any suggestions about this review? Currently I accept your opinion, remove deprecated and add javadoc about the behavior of `metadataKeys()`

Comment (langchain4j) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Sorry for delays, having too much on my plate lately, I will try to review it asap

Comment (Martin7-1) [reply langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStore.java]: Thank you so much!

**Comment by dliubarskyi** [comment]
@Martin7-1 I'm so sorry for this being dragged for so long! Perhaps you want to move redis module to the community repo to unblock these changes?

**Comment by Martin7-1** [comment]
@dliubarskyi No problem!

**Comment by Martin7-1** [comment]
Close as redis has been migrated to community repo.
